### PR TITLE
Improve the ergonomics of creating local buffers

### PR DESCRIPTION
**Pull request by nathansobo**
This PR renames `language::Buffer::new` to `language::Buffer::local` and simplifies its interface. Instead of taking a replica id (which should always be 0 for the local case) and a `BufferId`, which was awkward and verbose to construct, it simply takes text and a `cx`.

It uses the `cx` to derive a `BufferId` from the `EntityId` associated with the `cx`, which should always be positive based on the following analysis...

We convert the entity id to a u64 using this method on `EntityId`, which is defined by macros in the `slotmap` crate:

```rust
    pub fn as_ffi(self) -> u64 {
        (u64::from(self.version.get()) << 32) | u64::from(self.idx)
    }
```

If you look at the type of `version` in `KeyData`, it is non-zero:

```rust
#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct KeyData {
    idx: u32,
    version: NonZeroU32,
}
```

This commit also adds `Context::reserve_model` and `Context::insert_model` to determine a model's entity ID before it is created, which we need in order to assign a `BufferId` in the background when loading a buffer asynchronously.

Release Notes:

- N/A

### CODE REVIEW

File: crates/language/src/buffer.rs [hunk 1]
```diff
@@ -503,9 +503,9 @@ pub enum CharKind {
 
 impl Buffer {
     /// Create a new buffer with the given base text.
-    pub fn new<T: Into<String>>(replica_id: ReplicaId, id: BufferId, base_text: T) -> Self {
+    pub fn local<T: Into<String>>(base_text: T, cx: &mut ModelContext<Self>) -> Self {
```
Comment (rgbkrk) [inline crates/language/src/buffer.rs:506]: 👍

**Review by rgbkrk — APPROVED** [review]
This is so much cleaner. I appreciate the separate distinction from a remote buffer.

File: crates/gpui/src/gpui.rs [hunk 2]
```diff
@@ -165,6 +165,17 @@ pub trait Context {
         build_model: impl FnOnce(&mut ModelContext<'_, T>) -> T,
     ) -> Self::Result<Model<T>>;
 
+    /// Reserve a slot for a model to be inserted later.
+    /// The returned [Reservation] allows you to obtain the [EntityId] for the future model.
+    fn reserve_model<T: 'static>(&mut self) -> Self::Result<Reservation<T>>;
+
+    /// Insert a new model in the app context based on a [Reservation] previously obtained from [reserve_model].
```
Comment (osiewicz) [inline crates/gpui/src/gpui.rs:172]: This link is dead.
```suggestion
    /// Insert a new model in the app context based on a [Reservation] previously obtained from [`reserve_model`].
    ///
    /// [`reserve_model`]: Self::reserve_model
```

**Review by osiewicz — APPROVED** [review]
LGTM! This is indeed way easier on the eyes than `BufferId::new` galore we've started with.
My only concern is that `KeyData::as_ffi` is not guaranteed to be non-zero per `slotmap` API (even though the implementation gives us such a guarantee), but as long as we keep that in mind when upgrading `slotmap` version, we should be fine.

**Pull request merged**